server/src/main/java/org/folio/reservoir/server/UploadService.java

```diff
@@ -24,19 +29,67 @@
 
 public class UploadService {
 
+  private static final String LOCAL_ID = "localId";
   private static final Logger log = LogManager.getLogger(UploadService.class);
 
-  private Future<Void> uploadPayloadStream(ReadStream<JsonObject> upload,
-      IngestWriteStream ingestWriteStream) {
+  private Future<Void> uploadPayloadStream(Vertx vertx, ReadStream<JsonObject> upload,
+      IngestWriteStream ingestWriteStream, ModuleJsonPath jsonPath) {
     Promise<Void> promise = Promise.promise();
     if (ingestWriteStream != null) {
       ingestWriteStream.drainHandler(x -> upload.resume());
     }
     AtomicInteger number = new AtomicInteger();
     AtomicInteger errors = new AtomicInteger();
     upload.exceptionHandler(promise::tryFail);
-    upload.handler(r -> {
-      String localId = r.getString("localId");
+    upload.handler(r ->
+        validateIngestRecord(vertx, jsonPath, r, number, errors) 
+          .compose(rec -> {
+            if (rec == null || ingestWriteStream == null) {
+              return Future.succeededFuture(null);
+            }
+            if (ingestWriteStream.writeQueueFull()) {
+              upload.pause();
+            }
+            return ingestWriteStream.write(rec);
+          })
+          .onFailure(promise::tryFail)
+    );
+    upload.endHandler(e -> {
+      log.info("{} records processed", number.get());
+      promise.tryComplete();
+    });
+    return promise.future();
+  }
+
+  private static Future<Collection<String>> lookupPath(Vertx vertx, 
+      ModuleJsonPath jsonPath, JsonObject payload) {
+    return vertx.executeBlocking(p -> { 
+      Collection<String> strings = jsonPath.executeAsCollection(null, payload);
+      p.complete(strings);
+    }, 
+    false);
+  }
+
+  private Future<JsonObject> validateIngestRecord(Vertx vertx, ModuleJsonPath jsonPath,
```

Comment:
This method does a lot of different things. Can't it just be getLocalId ? .. and then make the another method to validateLocalId and then a final method for logging?

Reply:
It's all there but could be split into different methods, is all I say.

Reply:
Makes sense.